pkg/agent/route/route_linux.go

```diff
@@ -1246,29 +1234,41 @@ func (c *Client) AddClusterIPRoute(svcIP net.IP) error {
 	// Collect stale routes.
 	var staleRoutes []*netlink.Route
 	if curClusterIPCIDR != nil {
-		// If current destination CIDR is not nil, the route with current destination CIDR should be uninstalled.
+		// If the current ClusterIP CIDR exists, the route with the current ClusterIP CIDR should be uninstalled since
+		// a new route with a newly calculated ClusterIP CIDR has been installed.
 		route.Dst = curClusterIPCIDR
 		staleRoutes = []*netlink.Route{route}
 	} else {
-		// If current destination CIDR is nil, which means that Antrea Agent has just started, then all existing routes
-		// whose destination CIDR contains the first ClusterIP should be uninstalled, except the newly installed route.
-		// Note that, there may be multiple stale routes prior to this commit. When upgrading, all stale routes will be
-		// collected. After this commit, there will be only one stale route after Antrea Agent started.
 		routes, err := c.listIPRoutesOnGW()
 		if err != nil {
 			return fmt.Errorf("error listing ip routes: %w", err)
 		}
+		// If current ClusterIP CIDR doesn't exist, which means that Antrea Agent has just started, and the route for
+		// ClusterIP generated in the last round should be deleted.
 		for i := 0; i < len(routes); i++ {
-			if routes[i].Gw.Equal(gw) && !routes[i].Dst.IP.Equal(svcIP) && routes[i].Dst.Contains(svcIP) {
+			// Skip the route whose destination CIDR only has one address.
+			ones, _ := routes[i].Dst.Mask.Size()
+			if ones == ipv4AddrLength || ones == ipv6AddrLength {
+				continue
+			}
+			// Find stale route.
+			if routes[i].Dst.Contains(svcIP) && routes[i].Gw.Equal(gw) {
 				staleRoutes = append(staleRoutes, &routes[i])
+				// TODO: add a break here. Some old versions of Antrea have issue https://github.com/antrea-io/antrea/issues/3131.
+				//  There might be multiple stale routes and they should be removed too. Without the issue, there should
+				//  be only one stale route, and a break should be added here to avoid unnecessary iteration.
 			}
 		}
 	}
 
 	// Remove stale routes.
 	for _, rt := range staleRoutes {
 		if err = netlink.RouteDel(rt); err != nil {
-			return fmt.Errorf("failed to uninstall stale ClusterIP route %s: %w", rt.String(), err)
+			if err.Error() == "no such process" {
+				klog.InfoS("Failed to delete stale ClusterIP route since the route doesn't exist", "route", route)
```

Comment:
No need for this log if the routing entry does not exist?

Reply:
This was introduced in this PR https://github.com/antrea-io/antrea/pull/3827/files.